[BT] Add `Bettertransformer` support for FSMT

Sumanth077 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

younesbelkada reviewed (commented):
Thanks a lot for starting the PR! 
First of all, I can see several "unneeded" files (`.idea`, `*.xml` files) would you mind removing them as they are not needed at all 🙏 
Let me guide you here step by step: 

Step 1: I have already identified for you the name of the layer that needs to be replaced. For FSMT the layer is called [`EncoderLayer`](https://github.com/huggingface/transformers/blob/95754b47a6d4fbdad3440a45762531e8c471c528/src/transformers/models/fsmt/modeling_fsmt.py#L397)

Step 2: In [`optimum/bettertransformer/models/encoder_models.py`](https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/encoder_models.py) add a new class, here you can call it `FSMTEncoderLayerBetterTransformer` and follow the same template / logic as `BertEncoderLayerBetterTransformer` for assigning the new keys to the old one. The challenge here is to find out which key corresponds to exactly which part of the TransformerEncoderLayer. This step is described [here](https://huggingface.co/docs/optimum/bettertransformer/tutorials/contribute#step-2-building-the-xxxlayerbettertransformer-module) in the documentation

Step 3: Build the forward pass, again you can follow what is done in BERT and make sure to respect what is said [here](https://huggingface.co/docs/optimum/bettertransformer/tutorials/contribute#step-3-building-the-forward-pass) 

Step 4: Import you new class, so here `FSMTEncoderLayerBetterTransformer` inside [`optimum/bettertransformer/models/__init__.py`](https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/__init__.py), and add a new line inside `BETTER_TRANFORMER_LAYERS_MAPPING_DICT`, something that would look like: `"EncoderLayer":FSMTEncoderLayerBetterTransformer`

Step 5: You can check your conversion by doing something like:
```
from transformers import AutoModel
from optimum.bettertransformer import BetterTransformer

model = AutoModel.from_pretrained("hf-internal-testing/tiny-random-FSMTModel")
print(model)
```
And make sure your added layers are inside the model! 
Let me know if there is anything unclear

Sumanth077 commented:
Thankyou for the detailed breakdown @younesbelkada. I have followed and added all the things mentioned above.

1. Created a BertEncoderLayerBetterTransformer Class and assigned all the keys required.
2. Done Implementing the Forward Pass(Not sure whether this correct, but would love your detailed review and help here)
3. Imported the new class and added (EncoderLayer":FSMTEncoderLayerBetterTransformer) to the dictionary
4. Checked the conversion. The added layers are inside the model!
5. Finally removed the Unnecessary files.

Kindly review the recent commits @younesbelkada. Thankyou ❤️

younesbelkada commented:
Hi @Sumanth077 ! 
Thanks a lot for your great work and your message! For now it looks very nice to me and we're close to merging that! 
Now let's try to test your implementation before merging 💪 
Would you mind adding the line `"hf-internal-testing/tiny-random-FSMTModel"` exactly [here](https://github.com/huggingface/optimum/blob/5f398b3359ff09576855bb6487e7bb313a2816a0/tests/bettertransformer/test_bettertransformer_encoder.py#L47)
Then you can just run `pytest tests/bettertransformer//test_bettertransformer_encoder.py` and let me know how it goes!

younesbelkada commented on .gitignore:
```diff
@@ -134,4 +134,4 @@ dmypy.json
 
-.vscode
\ No newline at end of file
+.vscode.idea
```
Would you also like to remove this file as well 🙏

Sumanth077 commented:
Thankyou  @younesbelkada. Unfortunately Out of 14, 4 Test Cases got passed, 6 Test cases Failed and 4 got ignored.

Including a screenshot for reference:
<img width="1440" alt="Screenshot 2022-11-22 at 5 47 17 PM" src="https://user-images.githubusercontent.com/66694715/203311879-65e06f69-bc35-4d81-9018-c1dc101eca0c.png">

I will go through that. Also your assistance would be really great

younesbelkada commented:
Thanks! Could you please share with me the entire error trace?

Sumanth077 commented:
@younesbelkada Removed the .gitignorefile and also committed the remaining changes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum/pr_494). All of your documentation changes will be reflected on that endpoint.

younesbelkada commented:
I have merged #501 , could you merge your branch with the current main branch ? 🙏  Thanks!

younesbelkada commented on optimum/bettertransformer/models/encoder_models.py:
```diff
@@ -751,3 +751,111 @@ def forward(self, hidden_states, attention_mask, **__):
+                """
+        super().__init__(config)
+        # In_proj layer
+        self.in_proj_weight = nn.Parameter(
+            torch.cat(
+                [
+                    fsmt_layer.self_attn.self.q_proj.weight,
+                    fsmt_layer.self_attn.self.k_proj.weight,
+                    fsmt_layer.self_attn.self.v_proj.weight,
+                ]
+            )
+        )
+        self.in_proj_bias = nn.Parameter(
+            torch.cat(
+                [
+                    fsmt_layer.self_attn.self.q_proj.bias,
+                    fsmt_layer.self_attn.self.k_proj.bias,
+                    fsmt_layer.self_attn.self.v_proj.bias,
+                ]
+            )
+        )
```
```suggestion
        self.in_proj_weight = nn.Parameter(
            torch.cat(
                [
                    fsmt_layer.self_attn.q_proj.weight,
                    fsmt_layer.self_attn.k_proj.weight,
                    fsmt_layer.self_attn.v_proj.weight,
                ]
            )
        )
        self.in_proj_bias = nn.Parameter(
            torch.cat(
                [
                    fsmt_layer.self_attn.q_proj.bias,
                    fsmt_layer.self_attn.k_proj.bias,
                    fsmt_layer.self_attn.v_proj.bias,
                ]
            )
        )
```
This should fix some failing tests

younesbelkada commented on .gitignore:
```diff
@@ -1,137 +0,0 @@
-# Byte-compiled / optimized / DLL files
```
Sorry for the confusion, could you revert this file? What I meant before was to remove only the line that you have added!

Sumanth077 commented:
@younesbelkada I have done the suggested changes. Could you kindly confirm it. I can merge the branch if everything is fine.

younesbelkada commented on tests/bettertransformer/test_bettertransformer_encoder.py:
```diff
@@ -44,7 +44,7 @@
     "hf-internal-testing/tiny-random-Data2VecTextModel",
     "hf-internal-testing/tiny-random-MarkupLMModel",
     "hf-internal-testing/tiny-random-BertModel",
-    "ybelkada/random-tiny-BertGenerationModel",
```
Why this line has been removed?

younesbelkada commented on .gitignore:
```diff
@@ -132,6 +132,4 @@ dmypy.json
 # include small test model for tests
-!tests/assets/onnx/model.onnx
-
-.vscode
```
It seems that you have removed this line, could you make sure this file is the same as https://github.com/huggingface/optimum/blob/main/.gitignore ?

younesbelkada commented on optimum/bettertransformer/models/encoder_models.py:
```diff
@@ -751,3 +751,111 @@ def forward(self, hidden_states, attention_mask, **__):
+        )
+
+        # Out proj layer
+        self.out_proj_weight = fsmt_layer.self_attn.out_proj.dense.weight
+        self.out_proj_bias = fsmt_layer.self_attn.out_proj.dense.bias
+
+        # Linear layer 1
+        self.linear1_weight = fsmt_layer.fc1.weight
+        self.linear1_bias = fsmt_layer.fc1.bias
+
+        # Linear layer 2
+        self.linear2_weight = fsmt_layer.fc2.weight
+        self.linear2_bias = fsmt_layer.fc2.bias
+
+        # Layer norm 1
+        self.norm1_eps = fsmt_layer.self_attn_layer_norm.eps
+        self.norm1_weight = fsmt_layer.self_attn_layer_norm.weight
+        self.norm1_bias = fsmt_layer.self_attn_layer_norm.bias
+
+        # Layer norm 2
+        self.norm2_eps = fsmt_layer.final_layer_norm.eps
+        self.norm2_weight = fsmt_layer.final_layer_norm.weight
+        self.norm2_bias = fsmt_layer.final_layer_norm.bias
+
+        # Model hyper parameters
+        self.num_heads = fsmt_layer.self_attn.num_heads
+        self.embed_dim = fsmt_layer.self_attn.embed_dim
+
+        # Last step: set the last layer to `False` -> this will be set to `True` when converting the model
+        self.is_last_layer = False
```
It seems that the conversion is still not succesfull, you might need to review that once again. Could you make sure your conversion runs fine with the following script:

```
from transformers import AutoModel
from optimum.bettertransformer import BetterTransformer

model_id = "hf-internal-testing/tiny-random-FSMTModel"

hf_model = AutoModel.from_pretrained(model_id)
bt_model = BetterTransformer.transform(hf_model)
```

From there you will have different errors that you will need to dig a little bit and find out which key is not correct.

younesbelkada reviewed (commented):
Thanks a lot for your efforts! 

It seems that we still need to address few points before merging 💪 Keep it up! And let me know if you face into any issue

Sumanth077 commented:
@younesbelkada Sorry for the confusion, I have modified the gitignore file and as well as re added the line in "test_better transformer". 

And regarding the conversion, getting few errors after running the above script. I will go through and try to resolve that

younesbelkada commented:
Hi @Sumanth077 !
How is the integration going ? 💪 You are very close to merge this PR!  Let us know here if you face into any issue!

Sumanth077 commented:
Hi @younesbelkada I have fixed the changes and the conversion is fine.

Now I am not getting any errors while running the above script. 

Now only 4 test cases are failing, previously it was 6.

younesbelkada commented:
Could you let me know which test is failing? and push your latest changes here? Thanks!

Sumanth077 commented:
Hi @younesbelkada could you kindly help me in resolving the above Failed Test Cases. 

One issue I guess is in the defining of Forward Pass.

younesbelkada reviewed (approved):
Looking great! Thanks a lot for your clean implementation of `FSMT` integration for `BetterTransformer` ! 🎉 
Leaving the last review for @fxmarty and/or @michaelbenayoun

michaelbenayoun commented on tests/bettertransformer/testing_bettertransformer_utils.py:
```diff
@@ -70,7 +70,7 @@ def test_logits(self):
                 bt_hidden_states = converted_model(**inputs)[0]
 
-                if "gelu_new" in vars(random_config).values():
+                if "gelu_new" in list(random_config.to_dict().values()):
```
Great change, but no need to cast to `list` here!

younesbelkada replied:
thanks! but I think the `to_dict()` is necessary (otherwise I get a VERY weird error from `transformers`)

michaelbenayoun replied:
Yes exactly, but accessing the value is enough here

younesbelkada commented on tests/bettertransformer/testing_bettertransformer_utils.py:
```diff
@@ -70,7 +70,7 @@ def test_logits(self):
                 bt_hidden_states = converted_model(**inputs)[0]
 
-                if "gelu_new" in vars(random_config).values():
+                if "gelu_new" in list(random_config.to_dict().values()):
```
```suggestion
                if "gelu_new" in random_config.to_dict().values():
```

Sumanth077 commented:
Thankyou so much @younesbelkada. I couldn't have done it, without your Support. Really appreciate it❤️

Looking forward to contribute to more projects 🚀

younesbelkada commented:
Thanks a lot for your contribution @Sumanth077 ! 
One last thing, could you please add FSMT on the documentation together with the link to the paper: https://arxiv.org/abs/1907.06616 / the addition would go exactly [here](https://github.com/Sumanth077/optimum/blob/c5c13314a712ce734c1995630d31945ce0ad9641/docs/source/bettertransformer/overview.mdx?plain=1#L35) 
Thanks!

Sumanth077 commented:
Hi @younesbelkada I have mistakenly called git pull origin <local branch>, now instead of one change I done in doc file. It is asking me commit all these files again.

Could you help me with this.

<img width="1113" alt="Screenshot 2022-11-24 at 7 14 15 PM" src="https://user-images.githubusercontent.com/66694715/203799513-8967d0d0-5c76-472d-bd73-4810fabc7e35.png">

younesbelkada commented:
It is fine, I have just pushed it to the doc ;) Thanks again!
